rust/src/operations/create.rs

```diff
@@ -1,251 +1,367 @@
+    /// [dynamodb_lock::DynamoDbLockClient] describes additional options for the AWS atomic rename client.
+    pub fn with_storage_options(mut self, storage_options: HashMap<String, String>) -> Self {
+        self.storage_options = Some(storage_options);
+        self
     }
-}
 
-async fn do_create(
-    table_uri: String,
-    partition_id: usize,
-    mode: SaveMode,
-    metadata: DeltaTableMetaData,
-    protocol: Protocol,
-) -> DataFusionResult<SendableRecordBatchStream> {
-    let mut table = DeltaTableBuilder::from_uri(&table_uri)
-        .build()
-        .map_err(to_datafusion_err)?;
-
-    let actions = match table.load_version(0).await {
-        Err(_) => Ok(vec![
-            Action::protocol(protocol.clone()),
-            Action::metaData(MetaData::try_from(metadata.clone()).unwrap()),
-        ]),
-        Ok(_) => match mode {
-            SaveMode::Ignore => Ok(Vec::new()),
-            SaveMode::ErrorIfExists => {
-                Err(DeltaCommandError::TableAlreadyExists(table_uri.clone()))
-            }
-            _ => todo!("Write mode not implemented {:?}", mode),
-        },
+    /// Specify additional actions to be added to the commit.
+    pub fn with_actions(mut self, actions: impl IntoIterator<Item = Action>) -> Self {
+        self.actions.extend(actions);
+        self
+    }
+
+    /// Provide a [`DeltaObjectStore`] instance, that points at table location
+    pub fn with_object_store(mut self, object_store: Arc<DeltaObjectStore>) -> Self {
+        self.object_store = Some(object_store);
+        self
+    }
+
+    /// Consume self into uninitialized table with corresponding crate actions and operation meta
+    pub(crate) fn into_table_and_actions(
+        self,
+    ) -> DeltaResult<(DeltaTable, Vec<Action>, DeltaOperation)> {
+        if self
+            .actions
+            .iter()
+            .any(|a| matches!(a, Action::metaData(_)))
+        {
+            return Err(CreateError::MetadataSpecified.into());
+        }
+        if self.columns.is_empty() {
+            return Err(CreateError::MissingSchema.into());
+        }
+
+        let (table, storage_url) = if let Some(object_store) = self.object_store {
+            let storage_url = StorageUrl::parse(object_store.root_uri())?;
+            (
+                DeltaTable::new(object_store, Default::default()),
+                storage_url,
+            )
+        } else {
+            let storage_url =
+                StorageUrl::parse(self.location.ok_or(CreateError::MissingLocation)?)?;
+            (
+                DeltaTableBuilder::from_uri(&storage_url)
+                    .with_storage_options(self.storage_options.unwrap_or_default())
+                    .build()?,
+                storage_url,
+            )
+        };
+
+        // TODO configure more permissive versions based on configuration. Also how should this ideally be handled?
```

Comment:
We should write a function that looks at a list of actions and detects which features are used, then determine the protocol versions. It will be much easier if that logic is the responsibility of a single function than spread out.

I think for updates we only need to run that function on new actions and merge the result with the existing protocol versions on the table. There's probably also some concurrency resolution that needs to happen too, but I haven't yet thought about that part much..

Reply:
I might even say we should let clients set the protocol; we should consider that the responsibility of the library. What do you think? Are there use cases where they should set it?

Reply:
I think you are right - having the user choose the protocol and defaulting to "what spark uses" seems like the way to go. WE may in that case have to check option compatibility though?